File: client/src/music-client/shared/homeSectionOptions.ts
import { Range } from './index'
import { ArrayWithTotalCount, IdWithName } from '../../utils/types'
import { Track } from './track'

export enum TrackQueryFormType {
  MusicDbQuery = 'MusicDbQuery',
  YouTubeQuery = 'YouTubeQuery',
}

export interface HomeSectionOptions {
  tracklist: TracklistOptions
  tracklistShown: boolean
}

export type TracklistOptions = {
  query: {
    type: TrackQueryFormType.MusicDbQuery
    musicDbQuery: MusicDbTrackQueryParams
  } | {
    type: TrackQueryFormType.YouTubeQuery
    youTubeQuery: string
  }
  autoRefresh: boolean
  autoPlay: boolean
}

export interface HomeSectionPersistableState {
  options: HomeSectionOptions
  selectedTrackId?: number
  currentTrackId?: number
}

export interface HomeSectionPropsFromApi {
  options: HomeSectionOptions
  selectedTrackId?: string
  currentTrackId?: string
  tracksFromMusicDb?: ArrayWithTotalCount<Track>
  tracksFromYouTube?: Track[]
  youTubeChannels: IdWithName[]
  tags: string[]
}

export interface MusicDbTrackQueryParams {
  titleContains: string
  supportedYouTubeChannelsIds: string[]
  mustHaveEveryTag: string[]
  mustHaveAnyTag: string[]
  yearRange?: Partial<Range<number>>
  randomize: boolean
  relatedTracks: number[]
}

File: server/Music/DataAccess/Models/HomeSectionOptions.cs
using Music.Domain.Shared;

namespace Music.DataAccess.Models
{
    public class HomeSectionPersistableState
    {
        public class HomeSectionOptions
        {
            public class TracklistOptions
            {
                public class QueryForm_
                {
                    public TracksQuery MusicDbQuery { get; set; }

                    public string YoutubeQuery { get; set; }
                }

                public QueryForm_ QueryForm { get; set; }

                public bool AutoRefresh { get; set; }

                public bool AutoPlay { get; set; }
            }

            public TracklistOptions Tracklist { get; set; }

            public bool TracklistShown { get; set; }
        }

        public HomeSectionOptions Options { get; set; }

        public string SelectedTrackYoutubeId { get; set; }

        public string CurrentTrackYoutubeId { get; set; }
    }
}
